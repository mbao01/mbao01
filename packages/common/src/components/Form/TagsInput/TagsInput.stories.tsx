import type { Meta, StoryObj } from "@storybook/react";
import { TagsInput } from "./TagsInput";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: "Components/Form/TagsInput",
  component: TagsInput,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
    // backgrounds: [{ name: 'dark background', value: '#000', default: true }],
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  // decorators: [withTagsInput],
} satisfies Meta<typeof TagsInput>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    "aria-label": "Enter something",
  },
};

export const SmallTagsInput: Story = {
  args: {
    size: "sm",
    defaultTags: ["Hello", "World"],
    maxItems: 3,
    "aria-label": "Enter something",
  },
};

export const OutlineTagsInput: Story = {
  args: {
    outline: true,
    "aria-label": "Enter something",
  },
};

export const PrimaryTagsInput: Story = {
  args: {
    variant: "primary",
    "aria-label": "Enter something",
  },
};

export const DisabledTagsInput: Story = {
  args: {
    disabled: true,
    placeholder: "Enter something",
  },
};

export const WithPlaceholder: Story = {
  args: {
    variant: "info",
    placeholder: "Enter anything",
  },
};

export const WithDefaultTags: Story = {
  args: {
    defaultTags: ["Hello", "World"],
    "aria-label": "Enter something",
  },
};
