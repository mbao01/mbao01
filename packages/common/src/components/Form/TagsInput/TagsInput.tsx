"use client";

import type { ChangeEvent, ClipboardEvent, KeyboardEvent, MouseEvent, SyntheticEvent } from "react";
import { forwardRef, useCallback, useEffect, useRef, useState } from "react";
import { Cross2Icon } from "@radix-ui/react-icons";
import type { TagsInputProps } from "./types";
import { cn } from "../../../utilities";
import { Badge } from "../../Badge";
import { Input } from "../Input";
import {
  FORMATTING_REGEX,
  getTagClasses,
  getTagDeleteClasses,
  getTagsClasses,
  getTagsInputClasses,
  SPLITTER_REGEX,
} from "./constants";

export const TagsInput = forwardRef<HTMLDivElement, TagsInputProps>(
  (
    {
      onChange,
      defaultTags,
      placeholder,
      maxItems,
      minItems,
      className,
      dir,
      size,
      wide,
      outline,
      variant,
      disabled,
      "aria-label": ariaLabel,
      ...props
    },
    ref
  ) => {
    const inputRef = useRef<HTMLInputElement>(null);
    const [activeIndex, setActiveIndex] = useState(-1);
    const [inputValue, setInputValue] = useState("");
    const [disableInput, setDisableInput] = useState(false);
    const [disableButton, setDisableButton] = useState(false);
    const [isValueSelected, setIsValueSelected] = useState(false);
    const [selectedValue, setSelectedValue] = useState("");
    const [tags, setTags] = useState<string[]>(defaultTags ?? []);

    const parseMinItems = minItems ?? 0;
    const parseMaxItems = maxItems ?? Infinity;

    useEffect(() => {
      if (onChange) {
        onChange(tags, inputValue);
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [tags]);

    useEffect(() => {
      const verifyDisable = () => {
        if (tags.length - 1 >= parseMinItems) {
          setDisableButton(false);
        } else {
          setDisableButton(true);
        }
        if (tags.length + 1 <= parseMaxItems) {
          setDisableInput(false);
        } else {
          setDisableInput(true);
        }
      };
      verifyDisable();
    }, [tags, parseMinItems, parseMaxItems]);

    const onValueChangeHandler = useCallback(
      (tag: string) => {
        if (!tags.includes(tag) && tags.length < parseMaxItems) {
          setTags([...tags, tag]);
        }
      },
      [tags, parseMaxItems]
    );

    const removeValue = useCallback(
      (tag: string) => {
        if (tags.includes(tag) && tags.length > parseMinItems) {
          setTags([...tags.filter((item) => item !== tag)]);
        }
      },
      [tags, parseMinItems]
    );

    const handleSelect = useCallback(
      (e: SyntheticEvent<HTMLInputElement>) => {
        const target = e.currentTarget;
        const selection = target.value.substring(
          target.selectionStart ?? 0,
          target.selectionEnd ?? 0
        );

        setSelectedValue(selection);
        setIsValueSelected(selection === inputValue);
      },
      [inputValue]
    );

    const handlePaste = useCallback(
      (e: ClipboardEvent<HTMLInputElement>) => {
        e.preventDefault();
        const items = e.clipboardData.getData("text").split(SPLITTER_REGEX);
        const newTags = [...tags];
        items.forEach((item) => {
          const parsedItem = item.replaceAll(FORMATTING_REGEX, "").trim();
          if (
            parsedItem.length > 0 &&
            !newTags.includes(parsedItem) &&
            newTags.length < parseMaxItems
          ) {
            newTags.push(parsedItem);
          }
        });
        setTags(newTags);
        setInputValue("");
      },
      [tags, parseMaxItems]
    );

    const handleKeyDown = useCallback(
      (e: KeyboardEvent<HTMLInputElement>) => {
        const target = e.currentTarget;

        if (["ArrowLeft", "ArrowRight", "Backspace", "Delete", "Escape", "Enter"].includes(e.key)) {
          e.stopPropagation();

          const moveNext = () => {
            const nextIndex = activeIndex + 1 > tags.length - 1 ? -1 : activeIndex + 1;
            if (nextIndex === -1 && inputRef.current) {
              inputRef.current.focus();
              const inputLen = inputRef.current.value.length;
              inputRef.current.setSelectionRange(inputLen, inputLen);
            }
            setActiveIndex(nextIndex);
          };

          const movePrev = () => {
            const prevIndex = activeIndex - 1 < 0 ? tags.length - 1 : activeIndex - 1;
            setActiveIndex(prevIndex);
          };

          const moveCurrent = () => {
            const newIndex =
              activeIndex - 1 <= 0 ? (tags.length - 1 === 0 ? -1 : 0) : activeIndex - 1;
            setActiveIndex(newIndex);
          };

          switch (e.key) {
            case "ArrowLeft":
              if (dir === "rtl") {
                if (tags.length > 0 && activeIndex !== -1) {
                  moveNext();
                }
              } else {
                if (tags.length > 0 && target.selectionStart === 0) {
                  movePrev();
                }
              }
              break;

            case "ArrowRight":
              if (dir === "rtl") {
                if (tags.length > 0) {
                  movePrev();
                }
              } else {
                if (tags.length > 0 && activeIndex !== -1) {
                  moveNext();
                }
              }
              break;

            case "Backspace":
            case "Delete":
              if (tags.length > 0) {
                if (activeIndex !== -1 && activeIndex < tags.length) {
                  removeValue(tags[activeIndex]);
                  moveCurrent();
                } else {
                  if (target.selectionStart === 0) {
                    if (selectedValue === inputValue || isValueSelected) {
                      removeValue(tags[tags.length - 1]);
                    }
                  }
                }
              }
              break;

            case "Escape":
              setActiveIndex(activeIndex === -1 ? tags.length - 1 : -1);
              break;

            case "Enter":
              if (inputValue.trim() !== "") {
                e.preventDefault();
                onValueChangeHandler(inputValue);
                setInputValue("");
              }
              break;
          }
        }
      },
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [dir, activeIndex, tags, inputValue, removeValue]
    );

    const handleMouseDown = useCallback((e: MouseEvent) => {
      e.preventDefault();
      e.stopPropagation();
    }, []);

    const handleChange = useCallback((e: ChangeEvent<HTMLInputElement>) => {
      setInputValue(e.currentTarget.value);
    }, []);

    return (
      <div
        {...props}
        ref={ref}
        dir={dir}
        className={cn(getTagsClasses({ size, wide, outline, variant, disabled }), className)}
      >
        {tags.map((tag, index) => (
          <Badge
            tabIndex={activeIndex !== -1 ? 0 : activeIndex}
            key={tag}
            size={size}
            aria-disabled={disableButton}
            data-active={activeIndex === index}
            className={cn(getTagClasses({ size }), {
              "ring-1 ring-inset opacity-80": !disableButton && activeIndex === index,
            })}
            variant={variant === "default" ? "neutral" : variant}
          >
            <span className="truncate max-w-24">{tag}</span>
            <button
              type="button"
              aria-label={`Remove ${tag} option`}
              aria-roledescription="button to remove option"
              disabled={disableButton}
              onMouseDown={handleMouseDown}
              onClick={() => removeValue(tag)}
              className={cn(getTagDeleteClasses())}
            >
              <span className="sr-only">Remove {tag} option</span>
              <Cross2Icon className="h-4 w-4" />
            </button>
          </Badge>
        ))}
        <Input
          ref={inputRef}
          tabIndex={0}
          onSelect={handleSelect}
          aria-label={ariaLabel}
          disabled={disableInput}
          onKeyDown={handleKeyDown}
          onPaste={handlePaste}
          value={inputValue}
          onChange={activeIndex === -1 ? handleChange : undefined}
          placeholder={placeholder}
          onClick={() => setActiveIndex(-1)}
          className={cn(getTagsInputClasses({ size }), {
            "caret-transparent": activeIndex !== -1,
          })}
        />
      </div>
    );
  }
);

TagsInput.displayName = "TagsInput";
