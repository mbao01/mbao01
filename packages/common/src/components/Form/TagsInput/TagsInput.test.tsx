import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TagsInput } from "./TagsInput";

describe("TagsInput", () => {
  it("renders some default tags", () => {
    const { asFragment } = render(
      <TagsInput defaultTags={["Hello", "World"]} placeholder="Enter anything" />
    );

    expect(screen.getByRole("button", { name: "Remove Hello option" })).toBeVisible();
    expect(screen.getByRole("button", { name: "Remove World option" })).toBeVisible();
    expect(screen.getByPlaceholderText("Enter anything")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("and the user enters a tag", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<TagsInput aria-label="Enter anything" />);

    const tagsInput = screen.getByLabelText("Enter anything");

    await user.type(tagsInput, "Hello{Enter}");
    await user.type(tagsInput, "World{Enter}");

    expect(screen.getByRole("button", { name: "Remove Hello option" })).toBeVisible();
    expect(screen.getByRole("button", { name: "Remove World option" })).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("and the user can remove a tag", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<TagsInput aria-label="Enter anything" />);

    const tagsInput = screen.getByLabelText("Enter anything");

    await user.type(tagsInput, "Hello{Enter}");

    const helloTag = screen.getByRole("button", { name: "Remove Hello option" });

    await user.click(helloTag);

    expect(screen.queryByRole("button", { name: "Remove World option" })).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("and by using the keyboard the user can remove a tag", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<TagsInput aria-label="Enter anything" />);

    const tagsInput = screen.getByLabelText("Enter anything");

    await user.type(tagsInput, "Hello{Enter}");
    await user.type(tagsInput, "World{Enter}");
    await user.type(tagsInput, "Oh{Enter}");
    await user.type(tagsInput, "Little{Enter}");
    await user.type(tagsInput, "Dear{Enter}");

    await user.type(tagsInput, "{ArrowLeft}{Delete}");

    expect(screen.queryByRole("button", { name: "Remove Dear option" })).not.toBeInTheDocument();

    await user.type(tagsInput, "{Backspace}");

    expect(screen.queryByRole("button", { name: "Remove Little option" })).not.toBeInTheDocument();

    await user.type(tagsInput, "Hi{Backspace}{Backspace}{Backspace}");

    expect(screen.queryByRole("button", { name: "Remove Oh option" })).not.toBeInTheDocument();

    await user.type(
      tagsInput,
      "{ArrowLeft}{ArrowLeft}{ArrowLeft}{ArrowLeft}{ArrowLeft}{ArrowLeft}{Delete}"
    );

    expect(screen.queryByRole("button", { name: "Remove Hello option" })).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("and the user can only enter a certain minimum or maximum number of tags", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(
      <TagsInput minItems={1} maxItems={2} aria-label="Enter anything" />
    );

    const tagsInput = screen.getByLabelText("Enter anything");

    await user.type(tagsInput, "Hello{Enter}");

    const helloTag = screen.getByRole("button", { name: "Remove Hello option" });

    expect(helloTag).toBeDisabled();

    await user.type(tagsInput, "World{Enter}");

    expect(helloTag).toBeEnabled();
    expect(tagsInput).toBeDisabled();

    await user.click(helloTag);

    const worldTag = screen.getByRole("button", { name: "Remove World option" });

    expect(tagsInput).toBeEnabled();
    expect(worldTag).toBeDisabled();
    expect(helloTag).not.toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    ["has outline", true],
    ["has no outline", false],
  ] as const)("%s", (description, outline) => {
    const placeholder = `TagsInput ${description}`;
    const { asFragment } = render(<TagsInput outline={outline} placeholder={placeholder} />);

    expect(screen.getByPlaceholderText(placeholder)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    ["tiny", "xs"],
    ["small", "sm"],
    ["medium", "md"],
    ["large", "lg"],
  ] as const)("has %s (%s) size", (description, size) => {
    const placeholder = `TagsInput ${description}`;
    const { asFragment } = render(<TagsInput size={size} placeholder={placeholder} />);

    expect(screen.getByPlaceholderText(placeholder)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "default",
    "primary",
    "secondary",
    "accent",
    "ghost",
    "info",
    "success",
    "warning",
    "error",
  ] as const)("has %s variant", (variant) => {
    const placeholder = `${variant} badge`;
    const { asFragment } = render(
      <TagsInput key={variant} variant={variant} placeholder={placeholder} />
    );

    expect(screen.getByPlaceholderText(placeholder)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
