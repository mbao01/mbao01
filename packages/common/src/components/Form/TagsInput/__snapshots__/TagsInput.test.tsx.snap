// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TagsInput > and by using the keyboard the user can remove a tag 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <span
      aria-disabled="false"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none ring-1 ring-inset opacity-80"
      data-active="true"
      tabindex="0"
    >
      <span
        class="truncate max-w-24"
      >
        World
      </span>
      <button
        aria-label="Remove World option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove World option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <input
      aria-label="Enter anything"
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1 caret-transparent"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > and the user can only enter a certain minimum or maximum number of tags 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <span
      aria-disabled="true"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none"
      data-active="false"
      tabindex="-1"
    >
      <span
        class="truncate max-w-24"
      >
        World
      </span>
      <button
        aria-label="Remove World option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        disabled=""
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove World option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <input
      aria-label="Enter anything"
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > and the user can remove a tag 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <input
      aria-label="Enter anything"
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > and the user enters a tag 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <span
      aria-disabled="false"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none"
      data-active="false"
      tabindex="-1"
    >
      <span
        class="truncate max-w-24"
      >
        Hello
      </span>
      <button
        aria-label="Remove Hello option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove Hello option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <span
      aria-disabled="false"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none"
      data-active="false"
      tabindex="-1"
    >
      <span
        class="truncate max-w-24"
      >
        World
      </span>
      <button
        aria-label="Remove World option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove World option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <input
      aria-label="Enter anything"
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has accent variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-accent min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="accent badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has default variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border-0 min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="default badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has error variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-error min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="error badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has ghost variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-ghost min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="ghost badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has info variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-info min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="info badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has large (lg) size 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-16 px-5 text-lg"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1 h-6 text-lg"
      placeholder="TagsInput large"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has medium (md) size 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-12 px-4 text-sm"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1 h-5 text-sm"
      placeholder="TagsInput medium"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has no outline 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="TagsInput has no outline"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has outline 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border min-h-10 px-4 border-neutral-content"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="TagsInput has outline"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has primary variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-primary min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="primary badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has secondary variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-secondary min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="secondary badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has small (sm) size 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit rounded-md transition-all overflow-hidden min-h-8 px-3 py-1.5 text-sm"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1 h-4 text-sm"
      placeholder="TagsInput small"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has success variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-success min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="success badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has tiny (xs) size 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit rounded-md transition-all overflow-hidden min-h-6 px-2 py-1 text-xs"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1 h-3 text-xs"
      placeholder="TagsInput tiny"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > has warning variant 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden border border-warning min-h-10 px-4"
  >
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="warning badge"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TagsInput > renders some default tags 1`] = `
<DocumentFragment>
  <div
    class="flex items-center flex-wrap gap-1 h-fit py-2 rounded-md transition-all overflow-hidden min-h-10 px-4"
  >
    <span
      aria-disabled="false"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none"
      data-active="false"
      tabindex="-1"
    >
      <span
        class="truncate max-w-24"
      >
        Hello
      </span>
      <button
        aria-label="Remove Hello option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove Hello option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <span
      aria-disabled="false"
      class="badge badge-ghost text-xs rounded-md relative flex items-center gap-1 [&>span]:truncate [&>span]:max-w-24 aria-disabled:opacity-50 aria-disabled:cursor-not-allowed focus-visible:outline-none"
      data-active="false"
      tabindex="-1"
    >
      <span
        class="truncate max-w-24"
      >
        World
      </span>
      <button
        aria-label="Remove World option"
        aria-roledescription="button to remove option"
        class="transition-all duration-300 disabled:cursor-not-allowed [&:hover:not(:disabled)]:opacity-80"
        type="button"
      >
        <span
          class="sr-only"
        >
          Remove World option
        </span>
        <svg
          class="lucide lucide-x h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </button>
    </span>
    <input
      class="input rounded-md transition-all duration-100 min-h-fit outline-0 border-none h-5 min-w-fit flex-1 focus-visible:outline-0 focus-visible:ring-0 focus-visible:ring-offset-0 focus-visible:border-0 placeholder:text-muted-foreground px-1"
      placeholder="Enter anything"
      tabindex="0"
      value=""
    />
  </div>
</DocumentFragment>
`;
